### PR TITLE
Prepare the codebase to listen to the Storefront changes (1/4)

### PR DESCRIPTION
**Pull request by Juanpe**
Subtask of #1382 

### Description

This PR is the first step to preparing the codebase to respond to **Storefront** changes to update some stuff that is in the cache. 

Changes:
- Listen to Storefront changes for SK1 via `SKPaymentTransactionObserver` protocol using the `paymentQueueDidChangeStorefront` method.
- Listen to Storefront changes for SK2, creating a new listener (`StoreKit2TransactionListener `). 
- Update `PurchasesOrchestrator`, adding the new delegate methods.
- Fix failing tests related to these changes and add new ones.

### CODE REVIEW

File: Sources/Purchasing/PurchasesOrchestrator.swift
Comment (Juanpe) [inline Sources/Purchasing/PurchasesOrchestrator.swift:70]: In the future, we could create a single listener that would be composed of other "sublisteners"

File: Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift
Comment (Juanpe) [inline Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift:19]: I tried to add a test case to cover when the Storefront changes, but I couldn’t. Is there is a way? 🤔

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (NachoSoto) [inline Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:24]: I would implement `didSet` here like so:
```swift
didSet {
  if self.taskHandle !== oldValue {
    oldValue?.cancel()
  }
}
```

Then you only need to set it to `nil` or the new value.

Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:24]: much better 👍🏼 Thanks :)

Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:24]: Just a question, why do you have to check if they are identical, checking if they are not `equal` wouldn't be enough? 🤔

Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift]: Done! ✅  [cee546e](https://github.com/RevenueCat/purchases-ios/pull/1557/commits/cee546efcbb20d7163454f5a65a050db76ece160)

File: Sources/Purchasing/PurchasesOrchestrator.swift
Comment (NachoSoto) [reply Sources/Purchasing/PurchasesOrchestrator.swift:70]: Can this be `private(set)`?

File: Sources/Purchasing/PurchasesOrchestrator.swift
Comment (NachoSoto) [inline Sources/Purchasing/PurchasesOrchestrator.swift:73]: Ditto `private(set)`

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (NachoSoto) [inline Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:49]: This is duplicated, you can just do `self.taskHandle = nil` now :)
The advantage of what I suggested is that now it's impossible to have a reference to a cancelled task :)

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (NachoSoto) [inline Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:39]: Nit: our style guide requires explicit `self`:
```suggestion
        self.taskHandle = Task { [weak self] in
```

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (NachoSoto) [inline Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:42]: Maybe put this inside of `MainActor.run {}`?

File: Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift
Comment (NachoSoto) [reply Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift:19]: Yeah I don't think we can change the storefront mid-way through a test run unfortunately.

File: Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift
Comment (NachoSoto) [inline Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift:29]: I think you also need an `@available` annotation in the test too?

File: Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift
Comment (NachoSoto) [inline Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift:34]: You don't need the `!`. It's better if tests assume it's non-nil :)
```suggestion
        expect(self.listener.taskHandle).to(beNil())
```

File: Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift
Comment (NachoSoto) [inline Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift:43]: This is nice.

**Review by NachoSoto — APPROVED** [review]
Awesome!

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (NachoSoto) [inline Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:19]: Nit spaces:
```suggestion
protocol StoreKit2StorefrontListenerDelegate: AnyObject {

    func storefrontDidUpdate()
    
}
```

File: Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift
Comment (Juanpe) [reply Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift:29]: 🤔 I don't think so. [I've applied the `@available` attribute to the whole class](https://github.com/RevenueCat/purchases-ios/blob/cee546efcbb20d7163454f5a65a050db76ece160/Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift#L18) so I think it’s not needed, is it?

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:51]: Yeah! good catch! I forgot to update this line 🤦🏼‍♂️

File: Sources/Purchasing/PurchasesOrchestrator.swift
Comment (Juanpe) [reply Sources/Purchasing/PurchasesOrchestrator.swift:73]: same here 🙂

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift]: Done! ✅  ([171db44](https://github.com/RevenueCat/purchases-ios/pull/1557/commits/171db446d23bf8e2adfbf138cb65419f16333c07))

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:44]: Good point 👍🏼

Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift]: Done! ✅ ([ad5e799](https://github.com/RevenueCat/purchases-ios/pull/1557/commits/ad5e799ac4cf64a1e43076ab3b98b570f58c2e16))

File: Sources/Purchasing/PurchasesOrchestrator.swift
Comment (NachoSoto) [reply Sources/Purchasing/PurchasesOrchestrator.swift:70]: That's fine then!

Comment (NachoSoto) [reply Sources/Purchasing/PurchasesOrchestrator.swift:70]: That's fine then!

File: Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift
Comment (Juanpe) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:53]: I've reverted the change because when the listener is released, the `didSet` observer is not called 😅 so the task is not cancelled. [I've cancelled the task when the listener is released](https://github.com/RevenueCat/purchases-ios/blob/ad5e799ac4cf64a1e43076ab3b98b570f58c2e16/Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift#L52) and [modified the test cases](https://github.com/RevenueCat/purchases-ios/blob/ad5e799ac4cf64a1e43076ab3b98b570f58c2e16/Tests/StoreKitUnitTests/StoreKit2StorefrontListenerTests.swift#L29) to cover this scenario.

Comment (NachoSoto) [reply Sources/Purchasing/StoreKit2/StoreKit2StorefrontListener.swift:53]: Oooh that's a great catch I forgot!!

**Comment by NachoSoto** [comment]
I'll wait to see if @aboedo has any feedback before merging this.

**Comment by Juanpe** [comment]
Hi guys! 👋🏼  Did you have a chance to take a look at this PR? 🙂 The second PR is ready to be opened but I can't because I'm working on a fork and I can't select this branch as the base of the other PR. So, I have to wait for this PR to be merged to open the other one, otherwise, you would see the same changes 😅

**Review by taquitos — APPROVED** [review]
🎈🐐

**Pull request merged**
